### ISSUE TITLE
Namespace problems with Site-Wide Drush Commands

### ISSUE THREAD

**Issue by AdamPS** [drush-ops/drush#4641]
**Describe the bug**
Docs for [Site-Wide Drush Commands](https://docs.drush.org/en/9.x/commands/#site-wide-drush-commands) state to use a namespace `Drush\Commands`.  However this namespace isn't registered as an auto-loader, so it' doesn't work to try and use files in this directory.

**To Reproduce**
Create a site-wide command in `$PROJECT_ROOT/drush/Commands/ExampleCommands.php` that extends from a base class in the same directory.

**Expected behavior**
It works.

**Actual behavior**
The base class cannot be found.

**Workaround**
Add a line to your own composer.json.

```
  "autoload": {
    "psr-4": {
      "Drush\\Commands\\": "web/drush/Commands/"
    }
  },
```

### System Configuration
| Q               | A
| --------------- | ---
| Drush version?  | 10..3.6
| Drupal version? | Any
| PHP version     | 7.4.3
| OS?             | Linux

**Comment by AdamPS** [comment]
Proposed fix is to alter drush composer.json as per the workaround.  I can create a PR if you would accept this change.

**Comment by Sweetchuck** [comment]
I have been using site-wide Drush commands for a while, and definitely something wrong with the Command file detection.

But my workaround is the following:
File layout (relative from project root):
* `./drush/drush.yml`
* `./drush/drush.app.yml`
* `./drush/drush.local.yml`
* `./drush/Commands/contrib/foo/Commands/drush.yml`
* `./drush/Commands/contrib/foo/Commands/foo/FooBarCommands.php`

File content ./drush/drush.yml
```yaml
drush:
  paths:
    config:
      - '${drush.vendor-dir}/../drush/Commands/contrib/foo/Commands'
      - '${drush.vendor-dir}/../drush/drush.app.yml'
      - '${drush.vendor-dir}/../drush/drush.local.yml'
    include:
      - '${drush.vendor-dir}/../drush/Commands/contrib/foo'
```

File content ./drush/drush.app.yml
```yaml
options:
  root: '${drush.vendor-dir}/../docroot'
```

File content ./drush/Commands/contrib/foo/Commands/foo/FooBarCommands.php
```php
<?php

namespace Drush\Commands\foo;

class FooBarCommands {}
```

`alias d='vendor/bin/drush --config=drush @app.local'`
`d list --filter=foo`
> it works, but far from ideal.

**Comment by Sweetchuck** [comment]
Commands and Generator detection related issues:
* #4591 - symlinked drush/drush can't find the autoloader.php
* #4547 - Use the same rules to discover global Generators and site-wide Commands

**Comment by greg-1-anderson** [comment]
See the [Example Drush Extension](https://github.com/drush-ops/example-drush-extension). This example does not add an entry to `autoload` in its composer.json, but it could, similar to the suggestion in the OP, but in the extension composer.json rather than the site composer.json.

I don't think we should put an entry in Drush to autoload extension class files though.

**Comment by AdamPS** [comment]
It's fine by me to resolve this with documentation.  The two actions below would have resolved my uncertainty meaning I didn't raise an issue here.

1. Alter the composer.json of the example drush extension
2. Alter the [help page](https://docs.drush.org/en/9.x/commands/) for command authoring

**Comment by weitzman** [comment]
Pretty big improvement merged in #4696.
